includes/sanitizers/class-amp-style-sanitizer.php

```diff
@@ -1446,51 +1528,35 @@ function( $selector ) {
 		$stylesheet_set['processed_nodes'] = array();
 
 		$final_size = 0;
-		$dom        = $this->dom;
 		foreach ( $stylesheet_set['pending_stylesheets'] as &$pending_stylesheet ) {
 			$stylesheet = '';
 			foreach ( $pending_stylesheet['stylesheet'] as $stylesheet_part ) {
 				if ( is_string( $stylesheet_part ) ) {
 					$stylesheet .= $stylesheet_part;
 				} else {
 					list( $selectors_parsed, $declaration_block ) = $stylesheet_part;
-					if ( $should_tree_shake ) {
-						$selectors = array();
-						foreach ( $selectors_parsed as $selector => $parsed_selector ) {
-							$should_include = (
-								( $dynamic_selector_pattern && preg_match( $dynamic_selector_pattern, $selector ) )
-								||
-								(
-									// If all class names are used in the doc.
-									(
-										empty( $parsed_selector['classes'] )
-										||
-										0 === count( array_diff( $parsed_selector['classes'], $this->get_used_class_names() ) )
-									)
-									&&
-									// If all IDs are used in the doc.
-									(
-										empty( $parsed_selector['ids'] )
-										||
-										0 === count( array_filter( $parsed_selector['ids'], function( $id ) use ( $dom ) {
-											return ! $dom->getElementById( $id );
-										} ) )
-									)
-									&&
-									// If tag names are present in the doc.
-									(
-										empty( $parsed_selector['tags'] )
-										||
-										0 === count( array_diff( $parsed_selector['tags'], $this->get_used_tag_names() ) )
-									)
-								)
-							);
-							if ( $should_include ) {
+					$selectors                                    = array();
+					foreach ( $selectors_parsed as $selector => $parsed_selector ) {
+
+						// @todo Replacing with AMP component selectors should happen if $should_tree_shake as well.
+						if ( $should_tree_shake ) {
+							if ( $this->should_include_selector( $dynamic_selector_pattern, $parsed_selector, $selector ) ) {
 								$selectors[] = $selector;
 							}
+						} else {
+
+							// Lets map the amp elements and replace the correct one and then check if we should include the replacement.
+							$amp_selectors_parsed = $this->get_ampified_selectors_parsed( $parsed_selector, $selector );
+
+							// @todo Maybe we should also make sure here that the original selector is not needed?
+							if ( is_array( $amp_selectors_parsed ) && $this->should_include_selector( $dynamic_selector_pattern, $amp_selectors_parsed['parsed_selector'], $amp_selectors_parsed['selector'] ) ) {
+
+								// @todo Check if there could be selector's array?
+								$selectors[] = $amp_selectors_parsed['selector'];
+							} else {
+								$selectors = array_keys( $selectors_parsed );
+							}
```

Comment:
@westonruter Started adding the logic within this method since it seemed to be related with tree shaking in terms of adding/not adding the selector if it's not found after replacing. That's still WIP (e.g. currently works only in case `! $should_tree_shake` and is missing most of the mappings) but let me know if there are any general with this approach.

Reply:
Hmm, maybe one potential issue could be that the size of the style could exceed the limit if doing the replacement here.

Reply:
Moving the logic to `process_css_declaration_block` to avoid issues with style size changing after the size check.